### PR TITLE
NF: refactor syncFunction

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
Very short simplification, that I'll need for scoped storage, as sync will need to have a callback

### CODE REVIEW

**Comment by dae** [comment]
The main appeal of coroutines is to be able to execute asynchronous code without callbacks, so littering them throughout the new coroutine pathway feels like a large step in the wrong direction. Instead of passing a callback to a suspend fun, can't you just run the suspend fun, then run the callback after it? AFAICS only handleNewSync should need to change in Sync.kt.

File: AnkiDroid/src/main/java/com/ichi2/anki/Sync.kt [hunk 1]
```diff
@@ -136,18 +148,19 @@ private suspend fun handleNormalSync(
...
         }
 
         SyncCollectionResponse.ChangesRequired.FULL_UPLOAD -> {
-            handleUpload(deckPicker, auth, syncMedia)
+            handleUpload(deckPicker, auth, syncMedia, syncCallback)
         }
 
         SyncCollectionResponse.ChangesRequired.FULL_SYNC -> {
-            deckPicker.showSyncErrorDialog(SyncErrorDialog.DIALOG_SYNC_CONFLICT_RESOLUTION)
+            deckPicker.showSyncErrorDialog(SyncErrorDialog.DIALOG_SYNC_CONFLICT_RESOLUTION, syncCallback)
```
Comment (dae) [inline AnkiDroid/src/main/java/com/ichi2/anki/Sync.kt:163]: Does it make sense to call the callback before the user has had a chance to start the upload, download or cancel?

**Comment by Arthur-Milchior** [comment]
Right now, sync is still using AsyncTask.
In case you didn't know (personally, I learned last week-end), if we don't move to scoped storage by the end of the month, AnkiDroid is not listed on the play store anymore.
So my priority is to have something that just work and is not broken. And we can restart converting asynctask to coroutine later, and then potentially even removing callback, that'll be great.

That is, I fear I'm still not entirely clear about how things should work, even if we only used coroutines. Because if we start a normal sync, detect we must do a full sync, then we go back to UI, which itself starts a new task. 
Is something that is at the end of the initial task will still be executed?

I'm trying to follow the logic with stacktrace in the debugger, but the fact that I lose the external stack trace each time I enter a `launchCatchingTask` makes it hard to follow

File: AnkiDroid/src/main/java/com/ichi2/anki/Sync.kt (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/Sync.kt:163]: No. At least the callback I care about right now is something that is executed just after a succesfull sync (full or not), without leaving the user time to do anything on the UI in between

**Comment by Arthur-Milchior** [comment]
drafting. I may have a way to do the automated offer to upgrade without going through callback

**Comment by dae** [comment]
It would be possible to refactor showSyncErrorDialog() into an async version so that the chain of async calls only terminates after the user has chosen a direction and the full sync has completed (see userAcceptsSchemaChange for an example of an async dialog). But wouldn't it be easier to just check the sync status/badge status and warn the user if they haven't synced?

**Review by david-allison — CHANGES_REQUESTED** [review]
This is messy code, and it would be ideal to find another architecture which allows this to occur. First commit is fine (and in the code).
